### PR TITLE
Refactored Code Organization

### PR DESCRIPTION
**Pull request by david3214**
I moved a bunch of files around, and created a few extra folders. Find if things break.

@ComputationalPotato I'm not sure why but the database is reading in a bunch of blank employees even though the database.csv file has data. With the transition to a Data class I'm not sure how to actually view how the attributes are being set (or not set) This means you can't login to the app currently on this branch. I'm not sure if it is related to moving around the files, or the inital change to a dataclass

### CODE REVIEW

**Review by vla5qpb7 — APPROVED** [review]

**Comment by ComputationalPotato** [comment]
Is this both when running the exe and running main.py? If it's just the exe then I'd guess that it is related to how the datas in the .spec are being done differently. I put the database.csv in a database folder on my local branch and it works fine.

**Comment by david3214** [comment]
> Is this both when running the exe and running main.py? If it's just the exe then I'd guess that it is related to how the datas in the .spec are being done differently. I put the database.csv in a database folder on my local branch and it works fine.

I branched from master, so I don't have it running from main.py, but when I threw in print statements for the exe it looks like it read the file because it made a bunch of employee objects, but when I printed out their id's it was a list of empty strings. This leads me to thinking it's not due to the datas, because it was able to find all the files

**Comment by david3214** [comment]
If you build the debug.spec and run the exe from the terminal you can see any print statements you put in the code

**Comment by ComputationalPotato** [comment]
I'm using the database.csv in the database folder, and it works for me. Are the employees entirely blank or just most of their fields? Also, is the ini file set up correctly? The default settings for the ini file are for the same format as our current database.csv, and if there is no ini file, then a default one is created. Also, by default, you will be unable to log in as all of the employees in the database.csv don't have their passwords set('testing' branch has a patch for that).

**Comment by david3214** [comment]
That last statement is probably why, if they don't have passwords anymore then I would be able to log in. Although your other comments were helpful it would have been nice if you had checked out the branch and looked at the changes instead of making general guesses. In this branch the Database is under the resource tab, the ini file exists, and runs.  (If the database file didn't exist then we would have either gotten the file doesn't exist error or a crash)

**Comment by ComputationalPotato** [comment]
I looked at your commit and didn't notice anything that would be causing the issue you described, but if it was caused by switching to Dataclass then it was probably a bug that I fixed without even realizing on my branch. Looking through your commit was also difficult because of the huge number of files added making my browser have an aneurysm. The config file not being set correctly and causing employees to be blank is something that had happened to me before, and I fixed it by deleting the ini file. 

checking your database file, it appears to be the professor's employees.csv renamed to database.csv. Your ini file is using the default settings, which are for importing a file in the default format(exported by the project using the default settings).

**Comment by david3214** [comment]
I found the issue is that the empToContainer doesn't work for any fields that are named the same. For example, when reading a row from the database.csv the header for Emp ID is read as EmpID, which doesn't exist as a key in the empToContainer dictionary, meaning it doesn't get passed to the employee object

**Comment by david3214** [comment]
This should finally be good, will yall take a look and let me know what you think?

File: Config/Database.py [hunk 1]
```diff
@@ -1,27 +1,27 @@
...
 contToEmp = dict((empToContainer[k], k)for k in empToContainer)
 readcfg = configparser.ConfigParser()
 readcfg.optionxform = lambda optionstr: optionstr
 try:
-    with open('config.ini', 'r') as configfile:
+    with open(fetch_resource('./Resources/config.ini'), 'r') as configfile:
         readcfg.read_file(configfile)
 except (configparser.ParsingError, FileNotFoundError):
-    readcfg['TRANSLATION'] = empToContainer
+    readcfg['TRANSLATION'] = contToEmp
```
Comment (ComputationalPotato) [inline Config/Database.py:22]: using a config file like this causes the config file to not actually do anything, and cause `readcfg['TRANSLATION'].get(k.name, empToContainer[k.name])` to always use the fallback value, which is what it would get if the ini was created based on empToContainer. If this fixes a issue for you, then it is because your config.ini is set up incorrectly and should set to the default values from `readcfg['TRANSLATION'] = empToContainer`

File: Employee/Employee.py [hunk 2]
```diff
@@ -66,6 +66,8 @@ def __post_init__(self):
                 else:
                     val = t(self.__getattribute__(field.name))
                 setattr(self, field.name, val)
+        if self.hashed_password == '':
```
Comment (ComputationalPotato) [inline Employee/Employee.py:81]: this is a security vulnerability that was mentioned as fixed in the bug tracker. the database.csv on main already has the passwords set to their employee IDs

File: GUI/Components/Panels/AdvancedSearch.py [hunk 3]
```diff
@@ -83,7 +86,6 @@ def addEmployee(self):
 
     def searchAdvanced(self, *args):
         employees = DB.search(**self.getAllNonEmptyEntries())
-        self.master=typing.cast('Search',self.master)
```
Comment (ComputationalPotato) [inline GUI/Components/Panels/AdvancedSearch.py:86]: Why is this removed?

**Review by ComputationalPotato — CHANGES_REQUESTED** [review]
dist/debug and dist/main and build/ should be added to .gitignore, and everything inside them should be removed, 2037 files is insane. config.ini needs to be outside of the exe so that the user can use it.

File: Config/Database.py (hunk 1 above)
Comment (david3214) [reply Config/Database.py:22]: My config.ini file was setup by your lines of code, and we do want the translation to be going from container (database) to employee, because that is how you are using it later on in the code.

File: GUI/Components/Panels/AdvancedSearch.py (hunk 3 above)
Comment (david3214) [reply GUI/Components/Panels/AdvancedSearch.py:86]: Because the cast was already done in the init function, and self.master is never changed after that. No need to double cast it

File: Employee/Employee.py (hunk 2 above)
Comment (david3214) [reply Employee/Employee.py:81]: I never saw that bug tracker, so I'm not sure what you are referring to. The reason I added it in is currently there is no way to add a password via the application to any new employees or imported ones. This means when I tried to add a new user for the professor to use it never had a password set. I understand this could be a security violation, but please understand this is an example or prototype application. If this application were to be released we would need to change a lot of things, but right now there is no other way, which is why I added this line back in. In the future it definitely would be removed, but I don't think it's a good idea to remove it from our prototype right now

**Comment by david3214** [comment]
> dist/debug and dist/main and build/ should be added to .gitignore, and everything inside them should be removed, 2037 files is insane. config.ini needs to be outside of the exe so that the user can use it.

The user never needs to interact with the config.ini file, the program just needs to interface with it, which it is doing.

**Comment by ComputationalPotato** [comment]
> The user never needs to interact with the config.ini file, the program just needs to interface with it, which it is doing.

the whole point of the config.ini is so that the user can import a database that has different field names, such as the professor's file.

File: GUI/Components/Panels/AdvancedSearch.py (hunk 3 above)
Comment (ComputationalPotato) [reply GUI/Components/Panels/AdvancedSearch.py:86]: do you have typechecking turned on in vscode? for me, it still complained without the extra cast. If you have typechecking that can understand that the cast in the init affects the methods, pls tell me how you got it.

File: Config/Database.py (hunk 1 above)
Comment (ComputationalPotato) [reply Config/Database.py:22]: Your config.ini was setup with `readcfg['TRANSLATION'] = contToEmp`, I tested it. I just generated a new config.ini with `readcfg['TRANSLATION'] = empToContainer` and attached it to this(renamed it to a txt because github won't let me attach a .ini).
[config.txt](https://github.com/david3214/Team_2_CS_2450-601/files/10126339/config.txt)
the ini file is supposed to be setup in the format: [Employee_field] = [corresponding name from imported csv]
I tested your branch, the ini file doesn't work, changing a database header to a different name then changing the ini to match results in the field with the changed name not being read. I will update the code for the ini so that it will complain about missing fields instead of using a default value.

Comment (david3214) [reply Config/Database.py:22]: The dist/debug and dist/main were added to gitignore, as well as I fixed an issue causing build to still be tracked since it had already been added previously

**Comment by david3214** [comment]
> > The user never needs to interact with the config.ini file, the program just needs to interface with it, which it is doing.
> 
> the whole point of the config.ini is so that the user can import a database that has different field names, such as the professor's file.

That's great to understand. I didn't understand that since we had never discussed it and there were no comments surrounding it. Unfortunately, accessing files outside of the exe package is a huge pain, which is why we had to combine everything together into an included file in the installer. 

Also, in talking with the professor before the break he made it sound like he didn't need us to use his file specifically, but that it would handle importing similar databases. I know that is something that I initially pushed on because he hadn't made it clear until I asked more about it.

**Comment by ComputationalPotato** [comment]
Accessing files outside of the exe is easy, just use `open('config.ini','r'|'w')` as if the exe was main.py. just built and ran the exe and it created its own config.ini right next to it.

**Comment by david3214** [comment]
> Accessing files outside of the exe is easy, just use `open('config.ini','r'|'w')` as if the exe was main.py. just built and ran the exe and it created its own config.ini right next to it.

I didn't know that would work, interesting. I guess the issue was we couldn't do any relative paths based on the project itself. Also, if the exe file moves it no longer works, where as all the other files are contained within the exe

**Comment by david3214** [comment]
I just moved it to the config folder, but moved it to use open without fetch resource.

**Comment by ComputationalPotato** [comment]
I assumed that an installation would be of the normal format of:
the_folder
----the_executable.exe
----config.ini

**Review by ComputationalPotato — APPROVED** [review]

**Comment by david3214** [comment]
If you are going to make changes please at least correct it everywhere. You moved the config.ini file back to the resources folder, yet the code accesses in the config folder, making it useless there. The Resources folder is for accessing data using the fetch_resource command.... Right now it will just create a new config.ini file in the config folder.
